### PR TITLE
Optimize slice.{r}position result bounds check

### PR DESCRIPTION
**Pull request by arthurprs**
All the unrolling makes it hard for llvm to optimize it out.

Sample https://godbolt.org/g/R9nLvC

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]
Nicely done!

**Comment by dtolnay** [comment]
Travis is showing:

```
[00:58:51] error[E0599]: no method named `position` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:20:15
[00:58:51] |
[00:58:51] 20 | assert!(b.position(|&v| v == 9) == None);
[00:58:51] | ^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `position` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:21:15
[00:58:51] |
[00:58:51] 21 | assert!(b.position(|&v| v == 5) == Some(3));
[00:58:51] | ^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `position` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:22:15
[00:58:51] |
[00:58:51] 22 | assert!(b.position(|&v| v == 3) == Some(2));
[00:58:51] | ^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `position` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:23:15
[00:58:51] |
[00:58:51] 23 | assert!(b.position(|&v| v == 0) == None);
[00:58:51] | ^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `rposition` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:29:15
[00:58:51] |
[00:58:51] 29 | assert!(b.rposition(|&v| v == 9) == None);
[00:58:51] | ^^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `rposition` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:30:15
[00:58:51] |
[00:58:51] 30 | assert!(b.rposition(|&v| v == 5) == Some(4));
[00:58:51] | ^^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `rposition` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:31:15
[00:58:51] |
[00:58:51] 31 | assert!(b.rposition(|&v| v == 3) == Some(2));
[00:58:51] | ^^^^^^^^^
[00:58:51]
[00:58:51] error[E0599]: no method named `rposition` found for type `[{integer}; 5]` in the current scope
[00:58:51] --> /checkout/src/libcore/../libcore/tests/slice.rs:32:15
[00:58:51] |
[00:58:51] 32 | assert!(b.rposition(|&v| v == 0) == None);
[00:58:51] | ^^^^^^^^^
```

@bors r-

**Comment by arthurprs** [comment]
Sorry about that, broken tests... They're fixed now.

**Comment by kennytm** [comment]
@bors retry 

`dist-x86_64-musl`, failed to test stage2-libstd, #44899 ...? 

```
Testing libstd stage2 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-musl)
[01:00:44]    Compiling rand v0.0.0 (file:///checkout/src/librand)
[01:00:44]    Compiling alloc v0.0.0 (file:///checkout/src/liballoc)
[01:00:44]    Compiling std_unicode v0.0.0 (file:///checkout/src/libstd_unicode)
[01:00:44]    Compiling core v0.0.0 (file:///checkout/src/libcore)
[01:00:53]    Compiling std v0.0.0 (file:///checkout/src/libstd)
[01:01:35] rustc: /checkout/src/llvm/lib/Analysis/ValueTracking.cpp:1594: void computeKnownBits(const llvm::Value*, llvm::APInt&, llvm::APInt&, unsigned int, const {anonymous}::Query&): Assertion `(KnownZero & KnownOne) == 0 && "Bits known to be one AND zero?"' failed.
[01:01:35] error: Could not compile `alloc`.
[01:01:35] warning: build failed, waiting for other jobs to finish...
[01:03:14] error: build failed
```

**Comment by kennytm** [comment]
Same error. It's legit then.

@arthurprs Could you figure out what triggers #44899 here, and workaround it?

**Comment by arthurprs** [comment]
I moved the assume to the end. Can we try again?

File: src/libcore/slice/mod.rs [hunk 1]
```diff
@@ -1200,28 +1200,40 @@ macro_rules! iterator {
...
                     index -= 1;
                     if predicate(elt) {
                         SearchWhile::Done(Some(index))
                     } else {
                         SearchWhile::Continue
                     }
-                })
+                });
+                if let Some(index) = result {
+                    unsafe { assume(index < self.len); }
```
Comment (bluss) [inline src/libcore/slice/mod.rs]: `self.len` seems wrong here.

**Comment by arthurprs** [comment]
I can't parse the error from travis logs. Is it #44899?

File: src/libcore/slice/mod.rs (hunk 1 above)
Comment (arthurprs) [reply src/libcore/slice/mod.rs]: Thanks. I ran out of caffeine :coffee:

**Comment by kennytm** [comment]
@arthurprs No it's not #44899. The error message is

```
[00:54:44] thread '<unnamed>' panicked at 'explicit panic', /checkout/src/liballoc/../liballoc/tests/slice.rs:1194:16
```

Here: https://github.com/arthurprs/rust/blob/95912629bb43ecde40f47936bbb8701f7bdd8d30/src/liballoc/tests/slice.rs#L1194

I've restarted the CI build to check if the error is spurious.

**Comment by arthurprs** [comment]
I've no idea how this PR may be interacting with that :disappointed:

**Comment by kennytm** [comment]
@arthurprs Could you verify if the current version will still produce a more optimized assembly? It seems `if let Some(index) = result { Some(index) } else { None }` is equivalent to a no-op.

**Comment by bluss** [comment]
I think that version was just to see if it was the assume or something else that caused problems.

Assume is dark magic of the higher schools, so I'm not surprised there are sparks. I think it's best to not use it unless absolutely necessary.

**Comment by arthurprs** [comment]
This is scary stuff, the assume there is really triggering some UB.

**Comment by bluss** [comment]
I can't see anything wrong with the assumption, though

**Comment by arthurprs** [comment]
That what I'm thinking. That usize on the stack gets reused for something else and ...

I pushed one more attempt.

**Comment by kennytm** [comment]
The latest attempt caused the compiler to crash with SIGILL 😆 

```
[00:17:56] error: Could not compile `core`.
[00:17:56] 
[00:17:56] Caused by:
[00:17:56]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name core /checkout/src/libcore/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=8ae352561556b99e -C extra-filename=-8ae352561556b99e --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps` (signal: 4, SIGILL: illegal instruction)
[00:17:56] warning: build failed, waiting for other jobs to finish...
[00:18:00] error: build failed
```

**Comment by arthurprs** [comment]
Ok, this is definitely not safe.

File: src/libcore/slice/mod.rs [hunk 2]
```diff
@@ -1199,12 +1199,14 @@ macro_rules! iterator {
...
                 where F: FnMut(Self::Item) -> bool,
             {
+                let len = self.len();
                 let mut index = 0;
                 self.search_while(None, move |elt| {
                     if predicate(elt) {
                         SearchWhile::Done(Some(index))
                     } else {
                         index += 1;
+                        unsafe { assume(index < len); }
```
Comment (bluss) [inline src/libcore/slice/mod.rs]: (This one should be before += 1 to be correct.)

**Comment by bluss** [comment]
A bit of a tangential thought, maybe we should have a version of assume that's like debug_assert in debug builds and assume in release builds. I'm a proponent of debug checking almost all unsafe code.

File: src/libcore/slice/mod.rs (hunk 2 above)
Comment (arthurprs) [reply src/libcore/slice/mod.rs]: Good point.

Comment (kennytm) [reply src/libcore/slice/mod.rs]: @arthurprs  What if you make it `assume(0 < index && index <= len)`

Comment (arthurprs) [reply src/libcore/slice/mod.rs]: We can try that too. Lets try a few more things.

**Comment by scottmcm** [comment]
Is there a way to have a test for this?  I have a PR that will [conflict massively](https://github.com/rust-lang/rust/pull/45595/files#diff-1decd9f63c23ff5e439882aa3624d87aL1199) with this, and selfishly I'd like an easy way to ensure I don't regress it by accident.

Another thought: is this something that should be done for everything that's `TrustedLen`?

**Comment by arthurprs** [comment]
@kennytm is it possible to do a bors run but not merge this?

@scottmcm we're still trying to make this work🤔. As with most optimizations this is hard to test.

**Comment by kennytm** [comment]
@arthurprs Edit the `.travis.yml` and comment out any of the [`if: branch = auto` line](https://github.com/rust-lang/rust/blob/master/.travis.yml#L155) to test on the CI.

**Comment by arthurprs** [comment]
It failed in both musl builds :sad:  Any ideas?

`assume` is probably dangerous across function boundaries.

**Comment by arthurprs** [comment]
I'm gonna wait for https://github.com/rust-lang/rust/pull/45595 to get in so I can try more combinations.

**Comment by kennytm** [comment]
@arthurprs Have you tried to reproduce locally via Docker? Just run

```
./src/ci/docker/run.sh dist-x86_64-musl
```

**Comment by arthurprs** [comment]
I gave up trying, I keep getting

```
 ---> bec314029591
Successfully built bec314029591
Successfully tagged rust-ci:latest
configure: processing command line
configure: 
configure: rust.dist-src        := False
configure: build.extended       := True
configure: build.target         := ['x86_64-unknown-linux-musl']
configure: build.submodules     := False
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: build.locked-deps    := True
configure: llvm.ccache          := sccache
configure: build.openssl-static := True
configure: target.x86_64-unknown-linux-musl.musl-root := /musl-x86_64
configure: build.configure-args := ['--target=x86_64-unknown-linux-musl', '--musl ...
configure: 
configure: writing `config.toml` in current directory
Traceback (most recent call last):
  File "/checkout/src/bootstrap/configure.py", line 422, in <module>
    with open('Makefile', 'w') as f:
IOError: [Errno 13] Permission denied: 'Makefile
```

**Comment by dtolnay** [comment]
I am closing the PR because it doesn't look like it is being worked on. I filed #45964 to follow up and try again. It seems like `assume(index < len)` should have worked!

**Pull request closed without merging**